### DISCUSSION TITLE
How is TOML different than Windows INI files?

### DISCUSSION THREAD

**Discussion by Steven-Burns** [toml-lang/toml#845] [Q&A] [answered]
MS-DOS and Windows have had [INI files](https://en.wikipedia.org/wiki/INI_file) for more than 30 years, which TOML looks very similar.  How is TOML different?

**Comment by pradyunsg** [comment, marked as answer]
This is covered in the README. See the final paragraph of https://github.com/toml-lang/toml#comparison-with-other-formats.

**Comment by marzer** [comment]
TOML is deliberately similar to INI files, so much so that a sufficiently simple TOML file would be indistinguishable from an INI file.

The main difference is that INI is not a formal standard, so different implementations have different quirks, support different data types, etc. TOML is a standard.

**Reply by ChristianSi** [reply]
Because all string values in TOML need to be quoted, that is true only of files that don't contain *any* string values. Not your typical file, I'd say.

**Reply by marzer** [reply]
Note I said "sufficiently simple", not "typical".

Also I've encountered INI parsers that require (or at least support) quoted strings.

That's the point: INI is an informal mess.

**Reply by Sama-Developer** [reply]
[quote] TOML is a standard. [/quote]

I'm sorry, it's not.
It has a formal specification yes. But it's not a standard.
[https://en.wikipedia.org/wiki/TOML](https://en.wikipedia.org/wiki/TOML)

**Reply by marzer** [reply]
Semantics. It's correct English to call it a standard. I'm not saying there's something like an ISO body for it, I'm saying that there exists an official specification that should be regarded as the standard by which implementations should be measured. If you want to argue with that I don't care to waste time on you.

**Comment by LongTengDao** [comment]
Does `.ini` support **dotted key**?

**Reply by marzer** [reply]
Some implementations might? It's impossible to answer definitively since it's not standardised.

**Reply by ChristianSi** [reply]
Table names and keys may include dots, but they have no special meaning.

**Comment by ChristianSi** [comment]
Some differences:

* There is no standardized INI format, but various dialects.
* INI has no logical support for nesting beyond a single level (key/value pairs in tables). Table names and keys may include dots, but they have no special meaning.
* INI doesn't have inline arrays nor inline tables.
* INI doesn't have arrays of tables.
* It has no standardized syntax for multi-line strings.
* It contains only untyped data (whose interpretation is up to the parsing app), while TOML has strings, numbers, booleans, and various date/time related types.

**Comment by TimeSeriesLord** [comment]
TOML is just some guy named Tom trying to get famous with a specification for a file config format.

Wheel re-invention number 10 million 453 thousand 200 and 9.

Meanwhile, Sassenrath with his REBOL had the right idea long ago.

**Reply by izrik** [reply]
I assure you, he's famous for more than just a config file format :laughing:
